[WIP] Fluids and Alloy Smelter

Rover656 opened the pull request:
# Description

Adds EnderIO's fluids (basic, no behaviours as of yet) and the alloy smelter machine (alongside a bunch of supporting code).

Fixes issue(s): # (issue)

# Todo <!-- Remove this section if you're submitting an already-complete PR -->

Too many things to count right now :) This PR is soley for continuous reviews during the implementation process.

# Checklist:

- [x] My code follows the style guidelines of this project (.editorconfig, most IDEs will use this for you)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

<!-- Thanks to: https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/ for the building blocks of this template -->

Rover656 committed:
First pass on fluids

Rover656 committed:
Merge branch 'dev/1.18.x' into fluids-and-alloys

Rover656 committed:
Fix

Rover656 committed:
WIP, merging in upstream

Rover656 committed:
Merge branch 'dev/1.18.x' into fluids-and-alloys

Rover656 committed:
Alloy smelting works (without energy).

This is INCREDIBLY wip and anything can change. This is the first working implementation of the alloy smelter machine and all underlying pieces. Lots still to do.

Rover656 committed:
Merge branch 'dev/1.18.x' into fluids-and-alloys

Rover656 committed:
Capacitors and configs :)

Rover656 commented on enderio-base/src/main/java/com/enderio/base/common/item/tool/LevitationStaffItem.java:
```diff
@@ -26,20 +40,50 @@ protected int getMaxEnergy() {
+    @Nullable
+    @Override
+    public MultiCapabilityProvider initCapabilities(ItemStack stack, @Nullable CompoundTag nbt, MultiCapabilityProvider provider) {
+        // TODO: OUR OWN TAG
```
This todo is obsolete

Rover656 commented on enderio-base/src/main/java/com/enderio/base/common/util/UseOnly.java:
```diff
@@ -10,7 +10,7 @@
  * Marks a method to only be called on a side without using the {@link net.minecraftforge.api.distmarker.OnlyIn} Annotation
```
Need to update javadoc

justliliandev commented on enderio-base/src/generated/resources/data/minecraft/tags/fluids/water.json:
```diff
@@ -0,0 +1,25 @@
+{
```
remove once forge fluid PR is ready.
Only left as a note

justliliandev commented on enderio-base/src/main/java/com/enderio/base/common/item/tool/ColdFireIgniter.java:
```diff
@@ -103,6 +108,22 @@ public void appendHoverText(ItemStack stack, @Nullable Level level, List<Compone
+            pItems.add(is.copy());
+
+            getTankCap(is).ifPresent(handler -> {
+                if (handler instanceof AcceptingFluidItemHandler) {
```
use instanceof pattern matching

justliliandev replied:
also move pItems.add(is) into the if, because otherwise we would have 2 empty itemstacks

justliliandev commented on enderio-base/src/main/java/com/enderio/base/common/item/tool/LevitationStaffItem.java:
```diff
@@ -26,20 +40,50 @@ protected int getMaxEnergy() {
+    @Override
+    protected void onTickWhenActive(Player player, @Nonnull ItemStack pStack, @Nonnull Level pLevel, @Nonnull Entity pEntity, int pSlotId,
+        boolean pIsSelected) {
         player.addEffect(new MobEffectInstance(MobEffects.LEVITATION, 1)); // TODO: An upgrade to make it faster?
```
depending on empowerement level?

justliliandev commented on enderio-base/src/main/java/com/enderio/base/common/item/tool/LevitationStaffItem.java:
```diff
@@ -26,20 +40,50 @@ protected int getMaxEnergy() {
+        super.setFullCharge(pStack);
+        getTankCap(pStack).ifPresent(handler -> {
+            if (handler instanceof AcceptingFluidItemHandler) {
+                ((AcceptingFluidItemHandler) handler).setFluid(new FluidStack(EIOFluids.VAPOR_OF_LEVITY.get(), handler.getTankCapacity(0)));
```
use instanceof pattern matchin

justliliandev commented on enderio-base/src/main/java/com/enderio/base/common/item/tool/LevitationStaffItem.java:
```diff
@@ -26,20 +40,50 @@ protected int getMaxEnergy() {
+
+    @Override
+    public void renderOverlay(ItemStack pStack, int pXPosition, int pYPosition) {
+        ItemBarRenderer.renderFluidOverlay(pStack, 0, pXPosition, pYPosition);
```
missing super call for energy?

Rover656 replied:
Energy is handled with the vanilla durability bar

justliliandev commented on enderio-base/src/main/java/com/enderio/base/common/menu/SyncedMenu.java:
```diff
@@ -60,7 +60,7 @@ private void sync(boolean fullSync) {
         ListTag listNBT = new ListTag();
         for (int i = 0; i < clientToServerSlots.size(); i++) {
-            Optional<CompoundTag> optionalNBT = clientToServerSlots.get(0).toOptionalNBT();
+            Optional<CompoundTag> optionalNBT = clientToServerSlots.get(i).toOptionalNBT();
```
pls don't tell anyone, thank you

Rover656 replied:
Costed me about 30 mins debugging, dw

justliliandev commented on enderio-base/src/main/java/com/enderio/base/common/recipe/capacitor/CapacitorDataRecipe.java:
```diff
@@ -90,7 +92,7 @@ public ResourceLocation getId() {
         @Override
         public CapacitorDataRecipe fromJson(@Nonnull ResourceLocation recipeId, JsonObject json) {
-            Ingredient capacitor = Ingredient.fromJson(json.get("capacitor"));
+            Item capacitor = ForgeRegistries.ITEMS.getValue(new ResourceLocation(json.get("capacitor").getAsString())); // TODO: These may need more checks?
```
just check if capacitor is null or the items registry name is different to the one in the json. If so, throw an exception, that doesn't crash the game, but will be logged and the recipe is disgarded

justliliandev commented on enderio-base/src/main/java/com/enderio/base/common/recipe/capacitor/CapacitorDataRecipe.java:
```diff
@@ -99,20 +101,21 @@ public CapacitorDataRecipe fromJson(@Nonnull ResourceLocation recipeId, JsonObje
         @Nullable
         @Override
         public CapacitorDataRecipe fromNetwork(@Nonnull ResourceLocation recipeId, FriendlyByteBuf buffer) {
+            Item capacitor = ForgeRegistries.ITEMS.getValue(buffer.readResourceLocation());
```
fix read/write order. read resourcelocation after reading the cap data, or write resourcelocation first

Rover656 replied:
oopsie

justliliandev commented on enderio-machines/src/generated/resources/data/enderio_machines/recipes/alloy_smelting/vibrant_alloy_ingot.json:
```diff
@@ -0,0 +1,16 @@
+      "item": "enderio:energetic_alloy_ingot"
+    },
+    {
+      "item": "minecraft:ender_pearl"
```
use ForgeTag instead

justliliandev commented on enderio-machines/src/generated/resources/data/enderio_machines/recipes/alloy_smelting/redstone_alloy_ingot.json:
```diff
@@ -0,0 +1,16 @@
+  "type": "enderio_machines:alloy_smelting",
+  "ingredients": [
+    {
+      "item": "minecraft:redstone"
```
use Tag instead

justliliandev commented on enderio-machines/src/generated/resources/data/enderio_machines/recipes/alloy_smelting/redstone_alloy_ingot.json:
```diff
@@ -0,0 +1,16 @@
+      "item": "minecraft:redstone"
+    },
+    {
+      "item": "enderio:silicon"
```
use forge/silicon tag as it's used by refined storage (and probably others, also add our silicon to this tag)

justliliandev commented on enderio-machines/src/generated/resources/data/enderio_machines/recipes/alloy_smelting/pulsating_iron_ingot.json:
```diff
@@ -0,0 +1,16 @@
+      "tag": "forge:ingots/iron"
+    },
+    {
+      "item": "minecraft:ender_pearl"
```
use tag

justliliandev commented on enderio-machines/src/generated/resources/data/enderio_machines/recipes/alloy_smelting/energetic_alloy_ingot.json:
```diff
@@ -0,0 +1,19 @@
+  "type": "enderio_machines:alloy_smelting",
+  "ingredients": [
+    {
+      "item": "minecraft:redstone"
```
tag

justliliandev commented on enderio-machines/src/generated/resources/data/enderio_machines/recipes/alloy_smelting/energetic_alloy_ingot.json:
```diff
@@ -0,0 +1,19 @@
+      "tag": "forge:ingots/gold"
+    },
+    {
+      "item": "minecraft:glowstone_dust"
```
tag

justliliandev commented on enderio-machines/src/generated/resources/data/enderio_machines/recipes/alloy_smelting/end_steel_ingot.json:
```diff
@@ -0,0 +1,19 @@
+      "item": "enderio:dark_steel_ingot"
+    },
+    {
+      "item": "minecraft:obsidian"
```
use obsidian tag

justliliandev commented on enderio-machines/src/generated/resources/data/enderio_machines/recipes/alloy_smelting/end_steel_ingot.json:
```diff
@@ -0,0 +1,19 @@
+  "type": "enderio_machines:alloy_smelting",
+  "ingredients": [
+    {
+      "item": "minecraft:end_stone"
```
use end_stones tag

justliliandev commented on enderio-machines/src/generated/resources/data/enderio_machines/recipes/alloy_smelting/electrical_steel_ingot.json:
```diff
@@ -0,0 +1,19 @@
+      "tag": "forge:ingots/iron"
+    },
+    {
+      "item": "enderio:coal_powder"
```
use forge:dusts/coal

justliliandev commented on enderio-machines/src/generated/resources/data/enderio_machines/recipes/alloy_smelting/electrical_steel_ingot.json:
```diff
@@ -0,0 +1,19 @@
+      "item": "enderio:coal_powder"
+    },
+    {
+      "item": "enderio:silicon"
```
use tag: see below

justliliandev commented on enderio-machines/src/generated/resources/data/enderio_machines/recipes/alloy_smelting/dark_steel_ingot.json:
```diff
@@ -0,0 +1,19 @@
+      "tag": "forge:ingots/iron"
+    },
+    {
+      "item": "enderio:coal_powder"
```
tag

justliliandev commented on enderio-machines/src/generated/resources/data/enderio_machines/recipes/alloy_smelting/dark_steel_ingot.json:
```diff
@@ -0,0 +1,19 @@
+      "item": "enderio:coal_powder"
+    },
+    {
+      "item": "minecraft:obsidian"
```
tag

justliliandev commented on enderio-machines/src/generated/resources/data/enderio_machines/recipes/alloy_smelting/conductive_iron_ingot.json:
```diff
@@ -0,0 +1,16 @@
+  "type": "enderio_machines:alloy_smelting",
+  "ingredients": [
+    {
+      "item": "minecraft:redstone"
```
tag

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/base/PoweredMachineEntity.java:
```diff
@@ -0,0 +1,100 @@
+
+    // Helper methods for gui:
+    @UseOnly(LogicalSide.CLIENT)
+    public Vector2i guiGetEnergy() {
```
rename to getGuiEnergy and maybe use a record for more descriptive names

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/client/widget/EnergyWidget.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Override
+    public void renderButton(PoseStack pPoseStack, int pMouseX, int pMouseY, float pPartialTicks) {
+        Minecraft minecraft = Minecraft.getInstance();
+        RenderSystem.setShader(GameRenderer::getPositionTexShader);
+        RenderSystem.defaultBlendFunc();
+        RenderSystem.enableDepthTest();
+        RenderSystem.setShaderTexture(0, WIDGETS);
```
maybe move rendersystem setup to some kind of GuiHelper in base, as we use that there for the EIOScreen as well

Rover656 replied:
This can be done in a cleanup PR sometime

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/block/MachineBlocks.java:
```diff
@@ -10,25 +10,46 @@
-        .block("fluid_tank", props -> new MachineBlock(props, MachineBlockEntities.FLUID_TANK))
+    public static final BlockEntry<MachineBlock> SIMPLE_ALLOY_SMELTER = REGISTRATE
+        .block("simple_alloy_smelter", props -> new MachineBlock(props, MachineBlockEntities.SIMPLE_ALLOY_SMELTER))
         .properties(props -> props.strength(2.5f,8))
         .loot(LootTableUtils::copyNBT)
         .item()
         .tab(() -> EIOCreativeTabs.MACHINES)
         .build()
         .register();
 
-    public static final BlockEntry<MachineBlock> ENCHANTER = REGISTRATE
-        .block("enchanter", props -> new MachineBlock(props, MachineBlockEntities.ENCHANTER))
+    public static final BlockEntry<MachineBlock> ALLOY_SMELTER = REGISTRATE
+        .block("alloy_smelter", props -> new MachineBlock(props, MachineBlockEntities.ALLOY_SMELTER))
         .properties(props -> props.strength(2.5f,8))
+        .loot(LootTableUtils::copyNBT)
         .item()
         .tab(() -> EIOCreativeTabs.MACHINES)
         .build()
```
just as a suggestion: move simple, normal and enhanced MachineBlocks into a group or EnumMap or something, with a helper methods to create those groups. Could unclutter the registrationfiles a bit

justliliandev replied:
maybe not for alloy smelter, as it also has to incorporate a furnace, but could be useful for all other machines

Rover656 replied:
will do this in a future PR

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/AlloySmelterBlockEntity.java:
```diff
@@ -0,0 +1,189 @@
+        this.mode = defaultMode;
+
+        // This can be changed by the gui for the normal and enhanced machines.
+        add2WayDataSlot(new EnumDataSlot<>(this::getMode, this::setMode, SyncMode.GUI));
```
add check for tier, before adding dataslot so this can never happen, even if the client uses a custom build jar.

justliliandev replied:
or add a check in the setMode method

Rover656 replied:
The getMode() method ignores the custom mode if the tier is too low to be fair

justliliandev replied:
Yeah, but I would still add a check for it

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/FluidTankBlockEntity.java:
```diff
@@ -2,9 +2,13 @@
 import com.enderio.machines.common.blockentity.data.sidecontrol.item.ItemHandlerMaster;
+import com.enderio.machines.common.blockentity.data.sidecontrol.item.ItemSlotLayout;
 import com.enderio.machines.common.menu.FluidTankMenu;
+import com.sun.jna.platform.win32.WinUser;
```
remove import

Rover656 replied:
Where the bloody hell did that come from

justliliandev replied:
Exactly my thought

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/base/PoweredMachineEntity.java:
```diff
@@ -0,0 +1,100 @@
+    @Override
+    public <T> LazyOptional<T> getCapability(@NotNull Capability<T> cap, @Nullable Direction side) {
+        if (cap == CapabilityEnergy.ENERGY) {
+            return energyCap.cast(); // TODO: FUTURE: Sided access stuff.
```
for the comment. I'm pretty sure energy was able to be inserted everywhere

Rover656 replied:
I need to make it possible to disable sided access if the IOConfig sets a side to disabled

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/energy/MachineEnergyStorage.java:
```diff
@@ -0,0 +1,115 @@
+    protected int storedEnergy;
+
+    // TODO: Provide via constructor
+    protected int baseMaxStored = 10000;
```
rename baseMaxStored to capacity

Rover656 committed:
A bunch of work. Most importantly -> Machine Models!

Rover656 commented:
I know it doesnt build :P

Rover656 committed:
Enhanced Machines

Also
- Fixed build
- Optimised the models I added earlier
- Tidied some of the textures I had resized

Rover656 committed:
IO Overlays and fix tall drops

Rover656 committed:
Fluid tank (block rendering) still need to do item. Then we're refactoring to high hell

Rover656 committed:
Revert testing change

Rover656 committed:
Read desc

- Refactored some packages
- Fluid Tank now uses water colour from world
- Fluid tank item now displays fluid
- Refactored machine blockstate generation

Rover656 committed:
Enchanter model, more touchups, removed unused textures

justliliandev commented on enderio-base/src/main/java/com/enderio/base/client/gui/screens/CoordinateMenuScreen.java:
```diff
@@ -1,8 +1,8 @@
-package com.enderio.base.client.screen;
+package com.enderio.base.client.gui.screens;
```
screens -> screen, because everywhere else nouns are singular in package names

justliliandev commented on enderio-base/src/main/java/com/enderio/base/client/model/composite/CompositeBakedModel.java:
```diff
@@ -0,0 +1,83 @@
+
+    @Override
+    public TextureAtlasSprite getParticleIcon() {
+        if (particleSupplier != null)
```
if particleSupplier can be null mark the constructorparameter and field with @Nullable

justliliandev commented on enderio-base/src/main/java/com/enderio/base/client/model/composite/CompositeBakedModel.java:
```diff
@@ -0,0 +1,83 @@
+    public TextureAtlasSprite getParticleIcon() {
+        if (particleSupplier != null)
+            return particleSupplier.get();
+        return Minecraft.getInstance().getTextureAtlas(TextureAtlas.LOCATION_BLOCKS).apply(MissingTextureAtlasSprite.getLocation());
```
can this be extracted and moved to a different place, because this is also used in PaintedModels

justliliandev commented on enderio-base/src/main/java/com/enderio/base/client/model/composite/CompositeModelComponent.java:
```diff
@@ -0,0 +1,72 @@
+import net.minecraft.resources.ResourceLocation;
+
+public record CompositeModelComponent(ResourceLocation model, Vector3f translation, Vector3f rotation, boolean particleProvider) {
+    public JsonObject toJson() {
```
add space before method

justliliandev commented on enderio-base/src/main/java/com/enderio/base/client/model/composite/CompositeModelGeometry.java:
```diff
@@ -0,0 +1,63 @@
+            componentModels.add(baked);
+
+            // If we don't yet have the particle model, save
+            if (particleSupplier == null && baked != null && component.particleProvider()) {
```
Maybe log a warning if multiple particleprovider are selected?

justliliandev commented on enderio-base/src/main/java/com/enderio/base/client/model/composite/CompositeModelGeometry.java:
self reference or circular reference should cause an issue in the Bakery before. Vanilla doesn't have a protection against it and I don't think we need one either

justliliandev commented on enderio-base/src/main/java/com/enderio/base/common/item/tool/ColdFireIgniter.java:
```diff
@@ -103,6 +107,22 @@ public void appendHoverText(ItemStack stack, @Nullable Level level, List<Compone
+
+            getTankCap(is).ifPresent(handler -> {
+                if (handler instanceof AcceptingFluidItemHandler fluidHandler) {
+                    fluidHandler.setFluid(new FluidStack(EIOFluids.VAPOR_OF_LEVITY.get(), handler.getTankCapacity(0)));
```
maybe add one for every fluid in that tag?

Rover656 replied:
The fluid tag is unavailable at the time this is constructed apparently

justliliandev replied:
okay, is kinda weird, that if someone removes vapor of levity from the tag and it's filled with it but whatever

Rover656 replied:
its the creative tab anyway, I suppose its fine

Rover656 committed:
Final changes for this PR

justliliandev commented on enderio-machines/src/generated/resources/assets/enderio_machines/lang/en_us.json:
```diff
@@ -1,5 +1,13 @@
+  "block.enderio_machines.simple_alloy_smelter": "Simple Alloy Smelter",
+  "block.enderio_machines.simple_powered_furnace": "Simple Powered Furnace",
+  "gui.enderio.alloy_smelter.mode": "Smelting Mode",
+  "gui.enderio.alloy_smelter.mode_all": "All Smelting",
```
maybe rename to "alloying and smelting"? Smelting is tied to furnace in my brain, but maybe it's just me

Rover656 replied:
Agreed

Rover656 committed:
Missed the end steel retexture on here

justliliandev commented on enderio-base/src/main/java/com/enderio/base/client/renderer/RenderUtil.java:
```diff
@@ -0,0 +1,38 @@
+        case NORTH -> renderFace(pose, normal, consumer, texture, color, x, x + w, y + h, y, z, z, z, z, x, x + w, y, y + h);
+        case SOUTH -> renderFace(pose, normal, consumer, texture, color, x, x + w, y, y + h, 1.0f - z, 1.0f - z, 1.0f - z, 1.0f - z, x + w, x, y + h, y);
+        case WEST -> renderFace(pose, normal, consumer, texture, color, 1.0f - z, 1.0f - z, y + h, y, x, x + w, x + w, x, x, x + w, y, y + h);
+        case EAST -> renderFace(pose, normal, consumer, texture, color, z, z, y, y + h, x, x + w, x + w, x, x + w, x, y + h, y);
```
missing indent?

justliliandev commented on enderio-base/src/main/java/com/enderio/base/common/recipe/capacitor/CapacitorDataRecipe.java:
```diff
@@ -90,7 +91,9 @@ public ResourceLocation getId() {
-            Ingredient capacitor = Ingredient.fromJson(json.get("capacitor"));
+            Item capacitor = ForgeRegistries.ITEMS.getValue(new ResourceLocation(json.get("capacitor").getAsString())); // TODO: These may need more checks?
+            if (capacitor == null)
+                throw new RuntimeException("Capacitor in data recipe was not found!");
```
even though getValue is Nullable, a Forge Registry returns a default Value. For Items this is Items#AIR

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/client/MachinesClient.java:
```diff
@@ -0,0 +1,31 @@
+
+    @SubscribeEvent
+    public static void textureStitch(TextureStitchEvent.Pre event) {
+    }
```
remove empty event listener

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/client/MachinesClient.java:
```diff
@@ -0,0 +1,31 @@
+
+@Mod.EventBusSubscriber(value = Dist.CLIENT, bus = Mod.EventBusSubscriber.Bus.MOD)
+public class MachinesClient {
+    @SubscribeEvent
```
ad space between class and method

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/client/MachinesClient.java:
```diff
@@ -0,0 +1,31 @@
+import net.minecraftforge.fml.common.Mod;
+
+@Mod.EventBusSubscriber(value = Dist.CLIENT, bus = Mod.EventBusSubscriber.Bus.MOD)
+public class MachinesClient {
```
classname is imo not really descriptive

I would use something like MachinesClientEvents or MachineClientRegistryEvents

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/client/gui/screens/AlloySmelterScreen.java:
```diff
@@ -0,0 +1,68 @@
+import net.minecraft.world.entity.player.Inventory;
+
+public class AlloySmelterScreen extends EIOScreen<AlloySmelterMenu> {
+    private static final ResourceLocation BG_TEXTURE_SIMPLE_ALLOY = EIOMachines.loc("textures/gui/simple_alloy_smelter.png");
```
add empty line

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/client/gui/screens/AlloySmelterScreen.java:
```diff
@@ -0,0 +1,68 @@
+    @Override
+    protected void renderBg(PoseStack pPoseStack, float pPartialTick, int pMouseX, int pMouseY) {
+        super.renderBg(pPoseStack, pPartialTick, pMouseX, pMouseY);
+    }
```
remove empty override

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/client/gui/widget/EnergyWidget.java:
```diff
@@ -0,0 +1,75 @@
+        RenderSystem.setShader(GameRenderer::getPositionTexShader);
+        RenderSystem.defaultBlendFunc();
+        RenderSystem.enableDepthTest();
+        RenderSystem.setShaderTexture(0, WIDGETS);
```
move these kinda calls to a util in base?

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/client/gui/widget/ProgressWidget.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.function.Supplier;
+
+// Will need adapting for arrows, will do that as part of SAG mill
+public class ProgressWidget extends AbstractWidget {
```
Why AbstractWidget and not just the Widget interface? you wouldn't need to override all those mouseCalls
Note: I did it wrong too

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/client/gui/screens/AlloySmelterScreen.java:
```diff
@@ -0,0 +1,68 @@
+        super.init();
+        addRenderableWidget(new ProgressWidget(this, () -> menu.getBlockEntity().getProgress(), getGuiLeft() + 56, getGuiTop() + 36, 14, 14, 176, 0));
+        addRenderableWidget(new ProgressWidget(this, () -> menu.getBlockEntity().getProgress(), getGuiLeft() + 104, getGuiTop() + 36, 14, 14, 176, 0));
+
```
call addRenderableOnly for Widgets that don't need to listen to mouseglicks and keypresses

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/block/EnhancedMachineBlock.java:
```diff
@@ -0,0 +1,136 @@
+public class EnhancedMachineBlock extends ProgressMachineBlock {
+    public static final EnumProperty<DoubleBlockHalf> HALF = BlockStateProperties.DOUBLE_BLOCK_HALF;
+
+    private static final VoxelShape UPPER_SHAPE = Shapes.create(0, 0, 0, 1 , 0.64f, 1); //6.25
```
instead of comment just calculate that on the fly? 6.25f/16 is imo better than 0.64f
furthormore this comment doesn't make any sense, because 6.25f/16 is approx 0.39, so these comments kinda doesn't make any sense. I think you want 10.25f/16

Rover656 replied:
yea pretty sure i adjusted it after the comment, oopsie

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/block/SimpleMachineBlock.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.Random;
+
+public class SimpleMachineBlock extends ProgressMachineBlock implements Fallable {
+    public SimpleMachineBlock(Properties properties, BlockEntityEntry<? extends MachineBlockEntity> blockEntityType) {
```
add new line

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/block/ProgressMachineBlock.java:
```diff
@@ -0,0 +1,25 @@
+import net.minecraft.world.level.block.state.properties.BooleanProperty;
+
+public class ProgressMachineBlock extends MachineBlock {
+    public static final BooleanProperty POWERED = BlockStateProperties.POWERED;
```
add new line

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/AlloySmelterMode.java:
```diff
@@ -0,0 +1,55 @@
+    Alloys(true, false, MachineLang.ALLOY_SMELTER_MODE_ALLOY);
+
+    public final boolean canAlloy;
+    public final boolean canSmelt;
```
private and create getter to be the same as tooltip and pos

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/base/MachineBlockEntity.java:
```diff
@@ -32,25 +40,77 @@
+            if (!isServer()) {
+                modelData.setData(IO_CONFIG_PROPERTY, ioConfig); // TODO: Check if this is needed given its NBT Deserialized
+                requestModelDataUpdate();
+            }
```
it's needed, because it's a 2waydataslot. This method is called whenever the dataslot is deserialized, so also on the server

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/base/PoweredCraftingMachineEntity.java:
```diff
@@ -0,0 +1,223 @@
+            @Override
+            protected Float fromNBT(CompoundTag nbt) {
+                return nbt.getFloat("progress");
+            }
```
create an IntegerDataSlot instead of a anonymous inner class for this. That's much more resuseable

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/base/PoweredCraftingMachineEntity.java:
```diff
@@ -0,0 +1,223 @@
+
+                // We've tried loading the recipe now, forget it.
+                loadedRecipe = null;
+            }
```
extract that code into a function for better readability of the tick function

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/base/PoweredCraftingMachineEntity.java:
```diff
@@ -0,0 +1,223 @@
+    /**
+     * Get the recipe from the recipe manager.
+     */
+    private Optional<R> getRecipe() {
```
rename to findMatchingRecipe, as this method sounds similar to getCurrentRecipe.
Also make getCurrentRecipe public for TOP or HWYLA or WTHIT Integration

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/base/PoweredMachineEntity.java:
```diff
@@ -0,0 +1,98 @@
+    private final LazyOptional<IEnergyStorage> energyCap = LazyOptional.of(() -> this.energyStorage);
+
+    @UseOnly(LogicalSide.CLIENT)
+    private Vector2i clientEnergy;
```
create record for that, as capacity and current energy is just x and y with this

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/base/PoweredMachineEntity.java:
```diff
@@ -0,0 +1,98 @@
+        if (level.isClientSide) {
+            return clientEnergy;
+        }
+        return new Vector2i(energyStorage.getEnergyStored(), energyStorage.getMaxEnergyStored());
```
Log a warning when this method is called on the wrong side

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/base/PoweredMachineEntity.java:
```diff
@@ -0,0 +1,98 @@
+    public Optional<ICapacitorData> getCapacitorData() {
+        if (getSlotLayout().isPresent()) {
+            ItemSlotLayout layout = getSlotLayout().get();
+            int slot = layout.getFirst(ItemSlotLayout.SlotType.CAPACITOR);
```
maybe add a ItemSlotLayour#hasType(SlotType type) method, so the -1 check is not needed and the code is more readable? Or return an Optional

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/data/sidecontrol/item/ItemSlotLayout.java:
```diff
@@ -0,0 +1,137 @@
+        outputs.forEach(slot -> slotMap.put(slot, SlotType.OUTPUT));
+        upgradeSlots.forEach(slot -> slotMap.put(slot, SlotType.UPGRADE));
+        return new ItemSlotLayout(slotMap);
+    }
```
remove fromLegacy, as they are only used in not used methods

Rover656 replied:
aha some old migration code, oopsie

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/data/sidecontrol/item/ItemSlotLayout.java:
```diff
@@ -0,0 +1,137 @@
+        return slotTypeMap.entrySet().stream().filter(entry -> entry.getValue() == type).map(Map.Entry::getKey).collect(Collectors.toList());
+    }
+
+    public int getFirst(SlotType type) {
```
as metioned in another comment: change to Optional, so -1 checks are not needed

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/data/sidecontrol/item/ItemSlotLayout.java:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    public boolean slotIs(int slot, SlotType type) {
+        return slotTypeMap.getOrDefault(slot, SlotType.MISC) == type;
```
rename to isSlotType or something similar, as this method name is not intuitive (atleast to me)

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/data/sidecontrol/item/ItemSlotLayout.java:
```diff
@@ -0,0 +1,137 @@
+    private final int slotCount;
+
+    private ItemSlotLayout(Map<Integer, SlotType> slotTypeMap) {
+        this.slotTypeMap = slotTypeMap;
```
create ImmutableMap using Map#copyOf

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/energy/MachineEnergyStorage.java:
```diff
@@ -0,0 +1,121 @@
+
+    public int getMaxEnergyTransfer() {
+        return Math.round(baseMaxTransfer * capacitorSupplier.get().map(ICapacitorData::getBase).orElse(1.0f));
+    }
```
I've checked with 1.12 enderio and that didn't have any energy if no capacitor was present. So the correct orElse would be 0. My bad for suggesting the orElse(1.0f)

Rover656 replied:
no worries at all :P

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/entity/FallingMachineEntity.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Nullable
+    @Override
+    public ItemEntity spawnAtLocation(ItemStack stack, float offsetY) {
+        // Add block entity NBT to item stack
+        if (!stack.isEmpty() && blockData != null) {
+            CompoundTag itemNbt = new CompoundTag();
+            itemNbt.put("BlockEntityTag", blockData);
+            stack.setTag(itemNbt);
+        }
+        return super.spawnAtLocation(stack, offsetY);
+    }
```
it's a bit out of scope, but can u add that to PaintedSandEntity?

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/recipe/AlloySmeltingRecipe.java:
```diff
@@ -0,0 +1,167 @@
+
+    public AlloySmeltingRecipe(ResourceLocation id, NonNullList<Ingredient> ingredients, ItemStack result, int energy, float experience) {
+        if (ingredients.size() > 3) {
+            throw new IllegalArgumentException("Developer tried to create an invalid alloy smelting recipe!");
```
remove developer as the recipe can be created through a datapack

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/recipe/AlloySmeltingRecipe.java:
```diff
@@ -0,0 +1,167 @@
+            pBuffer.writeFloat(pRecipe.experience);
+        }
+
+        @Override
+        public void toJson(AlloySmeltingRecipe recipe, JsonObject json) {
+            JsonArray ingredients = new JsonArray(recipe.ingredients.size());
+            recipe.ingredients.forEach(ing -> ingredients.add(ing.toJson()));
+
+            json.add("ingredients", ingredients);
+
+            JsonObject jsonobject = new JsonObject();
+            jsonobject.addProperty("item", Registry.ITEM.getKey(recipe.result.getItem()).toString());
+            if (recipe.result.getCount() > 1) {
+                jsonobject.addProperty("count", recipe.result.getCount());
+            }
+
+            json.add("result", jsonobject);
+
+            json.addProperty("energy", recipe.energy);
+            json.addProperty("experience", recipe.experience);
+        }
```
move toJson up to the fromJson method

justliliandev commented on enderio-machines/src/main/java/com/enderio/machines/common/recipe/AlloySmeltingRecipe.java:
```diff
@@ -0,0 +1,167 @@
+            for (int i = 0; i < ingredientCount; i++) {
+                ingredients.set(i, Ingredient.fromNetwork(pBuffer));
+            }
+
```
use
```suggestion
pBuffer.readCollection(count -> NonNullList.withSize(count, Ingredient.EMPTY), Ingredient::fromNetwork);
```

justliliandev commented on enderio-machines/src/main/resources/assets/enderio_machines/models/block/dialing_device.json:
```diff
@@ -0,0 +1,87 @@
+{
+	"credit": "Made with Blockbench by Rover656",
```
remove credit, you already have it through git /j

SirEndii commented:
yay
time to smelt

Rover656 committed:
Process review comments

justliliandev commented on enderio-base/src/main/java/com/enderio/base/client/model/composite/CompositeModelGeometry.java:
```diff
@@ -49,6 +55,8 @@ public boolean isUvLocked() {
             if (particleSupplier == null && baked != null && component.particleProvider()) {
                 particleSupplier = baked::getParticleIcon;
+            } else if (particleSupplier != null && component.particleProvider()) {
+                EnderIO.LOGGER.warn("Duplicate particle supplier found on model " + modelLocation);
```
Duplicate -> Multiple

Rover656 committed:
Duplicate -> Multiple

flobernd commented:
Nice to see some work going on here :-)
